backend/api/Models/Concepts/ResearchFile/ResearchFileModel.cs

```diff
@@ -3,7 +3,7 @@
 
 namespace Pims.Api.Models.Concepts
 {
-    public class ResearchFileModel : FileModel
+    public class ResearchFileModel : FileModel<ResearchFilePropertyModel>
```

Comment:
not 100% sure templates is the answer here. If we are using inheritance, why not also create an interface for the FileProperty?
That way the list below will be something like:
```
public IList<FilePropertyModel> FileProperties { get; set; }
```

I think that way the upcasting to FileModel will work without any problems.

Reply:
Sorry, not sure what you mean by templates, are you referring to generics?

Reply:
Sorry, yeah. Same concept different name.

Reply:
The problem that I was running into is overriding FileProperties within the research file.
IList<ResearchFilePropertyModel> FileProperties
does not inherit from
IList<FilePropertyModel> FileProperties
even if ResearchFilePropertyModel extends FileProperties.
The best that I could get away with in that case was "hiding" the FileProperties field in the ResearchFileModel which generates a warning.

the reason that I need IList<ResearchFilePropertyModel> FileProperties in the research file is mostly for the mappings. I could try casting or maybe see if the mapping is smart enough to realize that a FileProperty is actually a ResearchFileProperty at runtime perhaps?

Reply:
You could explicitly cast which I think it will be safe. Assuming we are not going to be doing manual modifications to the object I think it would also be safe not to have the generic either.

This compiled and worked for me. 
```
using System.Collections.Generic;


public interface IFileProperty
{
    string a { get; set; }
}

public class RFileProperty : IFileProperty
{
    private string _a;

    public string a { get => _a; set => _a = value; }

    public string rOnly { get; set; }
}

public interface IFile
{
    IList<IFileProperty> Properties { get; }
}

public class RFile : IFile
{
    private IList<IFileProperty> _properties;

    public IList<IFileProperty> Properties { get => _properties; }

    public RFile()
    {
        _properties = new List<IFileProperty>();
    }

    public void AddProperty(RFileProperty property)
    {
        _properties.Add(property);
    }
}

class Test : IProblem
{
    public override void Run()
    {
        IFile rFile = new RFile();

        RFileProperty rProperty = new RFileProperty();
        rProperty.a = "some test";
        rProperty.rOnly = "only for R";

        rFile.Properties.Add(rProperty);

        foreach (IFileProperty prop in rFile.Properties)
        {
            RFileProperty rProp = (RFileProperty)prop;
            System.Console.WriteLine(rProp.a);
            System.Console.WriteLine(rProp.rOnly);
        }
    }
}
```

Reply:
I think the mapper can be specified to map between the dal and the explicit object. I am not sure how the serializer will handle that

Reply:
Sure, I'm willing to try it at least.

Reply:
Ok, this is a bit of a rabbit hole. Thinks work mostly fine. Mapster isn't smart enough to see the inheritence in the the nested list (even if you cast the type within the mapping). Not a big deal, can just do this:
```
.AfterMapping((src, dest) =>
                {
                    dest.FileProperties = new List<FilePropertyModel>();
                    foreach(var propertyResearchFile in src.PimsPropertyResearchFiles)
                    {
                        dest.FileProperties.Add(propertyResearchFile.Adapt<ResearchFilePropertyModel>());
                    }
                });
```

The bigger deal is the json serializer at the controller level. Outgoing it doesn't work out of the box (it only serializes the base classes properties). That can be worked around by manually serializing it using this and returning the result: https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-polymorphism

When receiving data we would need to do something similar. Manually parse the json as a stream and then use the steps from the above link to force it to parse the incoming json as the derived class.

So, it works, but in my mind this solution seems like more work then the above and possibly more error prone as well (if someone just parses a researchfile as a researchfile in the controller naively they wont get any of the fields in the derived property list).

What do you think?

Reply:
That's unfortunate that the JSON serializer is not working as we want. This is a model only for marshaling the file from and to the controller right? If that is the case maybe just specified mappings are necessary without any inheritance or generics. Otherwise, I fear we will start debugging obscure issues on the frontend when the serializer doesn't do what we want if we add new data